### ISSUE TITLE
Regression in playback of H264 YT videos, partly dependent on having VP9 as the preferred codec.

### ISSUE THREAD

**Issue by RJVB** [zaps166/QMPlay2#608]
A regression occurred between 23.02.5.05 and the current 23.02.05.32 : H264 YT content no longer shows the video stream.

For now I've unblocked this *partly* by making H264 the preferred again; VP9-only videos continue to play with that setting from what I can tell.

This video plays only audio when VP9 is the preferred codec:
https://www.youtube.com/watch?v=oRNjS0-RSNw

There are also videos that don't play video in the current version at all, e.g.
(https://www.youtube.com/watch?v=b5DJtDduAGo)

In the older version the video plays fine.

Initially I thought this was entirely related to having VP9 selected as the preferred codec but that appears to have been wrong. I also thought I was getting an error message related to the playback or rather detection failure, but that turned out to be generated when I stopped playback.

Sorry for the reporting mess, I continue to discover new aspects.

**Comment by zaps166** [comment]
Is it 6a2caa3fbae7dbc753b65fd07f137e08bcac6657?

**Comment by RJVB** [comment]
So, looks like it's easy to fix this regression ... and now you have some examples to check if the comment in the code you tried to remove is still valid (and maybe to figure out some more things about it) ;)

**Comment by RJVB** [comment]
FWIW, do you know at this location in `YouTube::getYouTubeVideo()` if the user requested an audio-only action? If so you can use that to decide whether to `continue` in case of a DASH stream (= when video is expected), or not (= when only audio is requested).

**Comment by zaps166** [comment]
It's not audio-only, but the problem is that is always chooses AAC and H.264 (720p 30 FPS) without https://github.com/zaps166/QMPlay2/commit/6a2caa3fbae7dbc753b65fd07f137e08bcac6657. Something has changed in yt-dlp and/or YouTube. Now the problem is why it doesn't play some VP9 videos?

**Comment by RJVB** [comment]
I didn't say that! VP9 content plays fine from what I can tell - though I'll admit that I never try beyond 720P.

From what I understand, #6a2caa3 was to fix an issue in playing or downloading only the audio so maybe

```
if (!audioOnly && protocol.contains("dash", Qt::CaseInsensitive)) continue;
``` 
would be an acceptable compromise for now?

**Comment by zaps166** [comment]
> I didn't say that!

But that's how it works for me :smile:

Example video: https://www.youtube.com/watch?v=LXb3EKWsInQ

1. Revert 6a2caa3fbae7dbc753b65fd07f137e08bcac6657
2. Select VP9
3. Play
4. Result: H.264 720p 30 FPS and AAC audio (single URL).

**Comment by RJVB** [comment]
OK, just trying that, and indeed I get H264 content...

Which is of course preferable over no video content when you know there!

BTW, there was actual H264 content that didn't show video content even if H264 was the preferred codec.

So the question is why the video gets dropped. If you can figure that out then maybe you can fall back to using the DASH stream (IIUC that's what the reverted code does)?

Or ...

```
if (!audioOnly && <UserPrefersH264> && protocol.contains("dash", Qt::CaseInsensitive)) continue;
```

**Comment by RJVB** [comment]
Hmmm, so VLC doesn't use FFmpeg for this? An FFmpeg bug that has been fixed yet?

**Comment by RJVB** [comment]
So this seems like a reasonable temporary workaround for me:

```
        const auto protocol = format["protocol"].toString();
        if (!audioOnly && m_preferredH264 && protocol.contains("dash", Qt::CaseInsensitive))
        {
            // Skip MP4 DASH, because it doesn't work properly (old comment, is it still valid?)
            continue;
        }
```

but it would be great if there were an automatic fallback to H264, or otherwise one that the user can activate.

From the looks of it, you could provide (or enable) a button for this only in case DASH content is being streamed; if the user misses the video content s/he could click on that button which sets `m_preferredH264` temporarily and relaunches the video one way or another. Just simply from the start, of course.

Would such an approach be possible? A menu command would work too, but the best way to make certain the user notices the option is to put the button in the video frame... (or a message with a keyboard shortcut).

**Comment by RJVB** [comment]
BTW, since when does VLC play YT videos?!

**Comment by zaps166** [comment]
> BTW, since when does VLC play YT videos?!

Just paste obtained stream URL :sweat_smile: VLC can play HTTPS.

---

Codec choosing should be fixed.

**Comment by RJVB** [comment]
> Codec choosing should be fixed.

In the meantime maybe some simple fast way to select the preferred codec then, like a little drop-down next to the resolution selector that always shows the current setting?

The only reason I don't revert to H264 is that that does indeed give a (theoretically) lesser audio quality, which is also why I'd like a permanent visual reminder of the active codec.
The video quality should be almost identical (just a bit more bandwidth required for H264).

**Comment by zaps166** [comment]
> The only reason I don't revert to H264 is that that does indeed give a (theoretically) lesser audio quality

H.264 is a video codec, not audio codec :smile: You can mix them as you wish (QMPlay2 chooses Opus whenever possible, AAC is chosen because currently Opus requires DASH).

---

> like a little drop-down next to the resolution selector

No. Moreover both VP9 and H264 are using MP4 DASH which is partially broken on FFmpeg. Only low resolution videos are available in pure HTTPS. This is what `yt-dlp` is giving me in JSON.

**Comment by RJVB** [comment]
>
>H.264 is a video codec, not audio codec :smile:

I know that, I'm talking from the user standpoint, where I observe that I have a choice between VP9 and H264. The former setting gives better quality (audio too) but can no longer be used to watch certain content. That content can still be watched by switching to H264 in code with the revert and the combined check above (!audioOnly && m_preferredH264 && format=="dash") .

>> like a little drop-down next to the resolution selector
>
>No. Moreover both VP9 and H264 are using MP4 DASH which is partially broken on FFmpeg. Only low resolution videos are available in pure HTTPS. This is what `yt-dlp` is giving me in JSON.

You'll hopefully agree that it would be nice to have a workaround to make all videos viewable again while FFmpeg is being fixed (and that fixed version becomes available), the yt-dlp people fix what needs to be fixed etc. A workaround that keeps everyone happy during that time, so not a straight revert of the commit above...

Besides, what's wrong with making the preferred codec choice a little easier to access? It's not like there is no place for a dropdown with space for a 4-letter code on the toolbar of the YT search widget...

**Comment by zaps166** [comment]
Could you test this?

<details>
<summary>Expand diff</summary>

```diff
diff --git a/src/modules/Extensions/YouTube.cpp b/src/modules/Extensions/YouTube.cpp
index 5c16ea18..c475a07c 100644
--- a/src/modules/Extensions/YouTube.cpp
+++ b/src/modules/Extensions/YouTube.cpp
@@ -1170,6 +1170,13 @@ QStringList YouTube::getYouTubeVideo(const QString &param, const QString &url, I
         if (format.isEmpty())
             continue;
 
+        const auto protocol = format["protocol"].toString();
+        if (protocol.contains("dash", Qt::CaseInsensitive))
+        {
+            if (format.contains("fragment_base_url"))
+                continue;
+        }
+
         const auto itag = format["format_id"].toString().toInt();
         const auto url = format["url"].toString();
         const auto ext = format["ext"].toString();

```

</details>

**Comment by RJVB** [comment]
>Could you test this?


Sure, but tomorrow (or rather, later today ... I was planning to try not to go to bed at 4:30 again ;) )

**Comment by RJVB** [comment]
Well, that seems to work pretty well too. I trust you know whether this will cause audio-only playback/download to regress again (but I still think there's no need to "be clever" and `continue` in case of `audioOnly` - supposing the fix for #592 was exhaustive).

For https://www.youtube.com/watch?v=oRNjS0-RSNw I now get H264, 720p@25Hz and Opus audio. That's good enough (and this video actually looks like it was filmed at a lower res).

For https://www.youtube.com/watch?v=-FVkej1NKMc, I get H264, 720p@50Hz and Opus audio.

Now the interesting thing: for all the other videos filmed in that same playlist (and filmed at the same time; https://www.youtube.com/watch?v=b5DJtDduAGo, https://www.youtube.com/watch?v=wWg2NuJRif4, https://www.youtube.com/watch?v=pWqFk0IEahA and https://www.youtube.com/watch?v=pWqFk0IEahA), I get H264 but only 640x360@25Hz, Opus audio. With my own hack from above they all gave the higher video resolution but AAC audio.

I'm requestion 720p@60Hz btw.

The way things currently works it looks like the VP9/H264 codec preference choice could actually be a quality choice if you do something like

```
+        const auto protocol = format["protocol"].toString();
+        if (protocol.contains("dash", Qt::CaseInsensitive))
+        {
+            if (m_preferredH264 || format.contains("fragment_base_url"))
+                continue;
+        }
+
```

Either way that setting now has an effect on the audio codec too so there's also something to say to add a preference for the audio codec and possibly one indicating a priority. That is, whether the preference for the one modality is important enough to allow dropping the other modality (video without audio, audio with video). All becomes very complex but in the end it may not be possible to come up with something automatic that caters to everyone.

**Comment by zaps166** [comment]
> BTW, do you have a suggestion for a simple filter function that allows me to dump the relevant information from the `format` JSON object when the `continue` path is triggered (simply via `qCWarning(youtube)`)? There's too much in there to see at a glance if maybe there are other criteria that could be used to fine-tune the decision process.

Try `yt-dlp "https://www.youtube.com/watch?v=oRNjS0-RSNw" -j > yt.json` and open `yt.json` in web browser (e.g. Firefox). This allows you to easy explore the data. You can see all `formats` and `protocol` and `fragment_base_url`.

> With my own hack from above they all gave the higher video resolution but AAC audio.

I guess it's stream which contains video and audio together.

Sample rate doesn't matter (Opus is always at 48Khz, so it has to be resampled first). AAC can be worse at the same bitrate (it highly depends on encoder and settings used by youtube). AAC is patented and might have lower audio quality at the same bitrate, so I use Opus whenever possible. 48 kHz is usually a default settings for OSes, so no resampling on your side (using default settings). For good quality use Deezer, Tidal, buy CD or Vinyl or... use torrents :smile: 

---

QMPlay2 and FFmpeg currently don't support DASH with multi-quality streams, so I have to reject formats with `fragment_base_url` (I see this when DAHS doesn't work).

---

```
+            if (m_preferredH264 || format.contains("fragment_base_url"))
+                continue;
```

This will lower H.264 resolution,  FPS, audio in other videos (without this it's possible to run 1080p60, with your 'hack' it's 720p30 and AAC).

**Comment by RJVB** [comment]
The OSes I use all use 44.1Khz as the default resolution; PulseAudio on Linux is supposed to switch to 48kHz automatically. On Mac I can control the DAC sample rate manually or let QMPlay2 do it. So resampling during playback is not an issue.
I use YouTube as a source for music I want to download but not buy because I know that way there is _some_ respect of copyright (and all that). The quality is usually good enough for what I use that kind of music for, but evidently I prefer to pick the version with the best theoretical quality.


> This will lower H.264 resolution, FPS, audio in other videos (without this it's possible to run 1080p60, with your 'hack' it's 720p30 and AAC).

How/why? That hack is what I used yesterday and as I said I got 720P H264 videos (maybe even @50Hz) for the ones listed above. Those same videos only give me 640x360@25Hz with your proposed solution. That's with VP9 as the preferred codec, I haven't yet tried what happens when I select H264 as the preferred codec. Should be the same thing as yesterday though, as the conditional expression is the same...

But let's rewind. Is it not possible to design an "auto" algorithm and setting that finds the best video and audio streams?

**Comment by zaps166** [comment]
> The OSes I use all use 44.1Khz as the default resolution;

I thought about default behavior, not default "set-by-user" :smile: IIRC Windows has 48kHz, PipeWire also has 48kHz (with possibility to change or auto-detect).

PulseAudio - please no XD! It was horrible: high CPU usage, high latency, crackling audio or other weird audio glitches... :rofl: Hopefully we have PipeWire (no issues with audio since I have it) :smile: 

Obviously, resampling it's not an issue, it's lossy audio anyway.

> I use YouTube as a source for music I want to download but not buy because I know that way there is some respect of copyright (and all that).

I don't understand: if you download it from YT without buying, it's it a violation of copyright, too? Maybe you mean download to RAM/cache to listen?

> How/why? That hack is what I used yesterday and as I said I got 720P H264 videos (maybe even @50Hz) for the ones listed above. Those same videos only give me 640x360@25Hz with your proposed solution.

Your video is 720P (instead of 1080P as in YT), but other videos which can run H264 1080P, will run 720P only. Try my favorite test video: https://www.youtube.com/watch?v=LXb3EKWsInQ. Your video plays to me at 25 FPS, not 50 FPS with your hack.

**Comment by RJVB** [comment]
>I thought about default behavior

That's what I meant too. I don't think MSWin <10 used 48kHz as its default; I don't have 11 and the machine with MSWin10 was upgraded from what was it, Win7.


>PulseAudio - please no XD! It was horrible: high CPU usage

It's what I have and manage just fine with for now. I'd want something else (or bypass it) on a Linux audio cont server to avoid it's either-44-or-48 design but for now I use Win10 for that.

> Maybe you mean download to RAM/cache to listen?

Let's say that. 

>Your video is 720P (instead of 1080P as in YT), but other videos which can run H264 1080P, will run 720P only. Try my favorite test video: https://www.youtube.com/watch?v=LXb3EKWsInQ. Your video plays to me at 25 FPS, not 50 FPS with your hack.

I'll need to do more testing but I'm pretty certain I got them @50Hz yesterday.

> It's also possible to play Opus instead of AAC, but both will download, one will play... 😄

Yeah, I thought about that. A priori that should be OK (I have the bandwidth) that could be a user-selectable option. Or a documented effect of an "auto-max" setting ;)
Question is how you'd handle this when downloading ... re-mux the data into a container containing only the played tracks (or a container containing all downloaded tracks)?

**Comment by zaps166** [comment]
> that could be a user-selectable option.

Too much in UI especially for a fallback for certain videos only. DASH should be supported instead, but it's not easy. Opus is selected by default, AAC can be selected manually for some videos where there's a problem with DASH.

> Question is how you'd handle this when downloading ... re-mux the data into a container containing only the played tracks (or a container containing all downloaded tracks)?

It'll just download and mux both... :smile: If you download audio-only, Opus will be used. QMPlay2 is also stupid about downloading - when you download audio from YouTube tab, it downloads audio. When you download audio from Playlist tab, it downloads both and mux audio only.

**Comment by RJVB** [comment]
>Too much in UI especially for a fallback for certain videos only.

This one would definitely be in the settings dialog, which isn't that complex as far as the extensions page is concerned.

>DASH should be supported instead, but it's not easy.

I can imagine, if FFmpeg doesn't do it properly already (not even the latest version?). But you seem to have found the right way to get the best quality; my additional hack is no longer required (and in fact causes certain of my test videos to be rejected entirely).

Now, I really have to get back to figuring out why VLC does VideoToolBox acceleration on my 10.9 system and QMPlay2 doesn't ;)

**Comment by zaps166** [comment]
> Now, I really have to get back to figuring out why VLC does VideoToolBox acceleration on my 10.9 system and QMPlay2 doesn't ;)

Maybe too old? :astonished: It's 2013!!

I checked it this month and it worked for various codecs on the newest macOS (x64 on M1 via Rosetta).
